pkg/nsx/services/firewall.go

```diff
@@ -432,60 +713,108 @@ func (service *SecurityPolicyService) buildRuleSrcGroup(obj *v1alpha1.SecurityPo
 		DisplayName: &ruleSrcGroupName,
 		Tags:        peerTags,
 	}
+
+	ruleSrcGroupCount := 0
+	var err error = nil
 	for i, peer := range sources {
-		service.updatePeerExpressions(obj, &peer, &ruleSrcGroup, i)
+		ruleSrcGroupCount += service.updatePeerExpressions(obj, &peer, &ruleSrcGroup, i)
 	}
+	if ruleSrcGroupCount > MaxGroupExpressions {
+		errorMsg := fmt.Sprintf("Total number of rule source group expressions:%d exceed NSX limit of %d", ruleSrcGroupCount, MaxGroupExpressions)
+		err = errors.New(errorMsg)
+		log.Error(err, "buildRuleSrcGroup")
+	}
+	log.V(1).Info(fmt.Sprintf("ruleSrcGroupCount: %v in buildRuleSrcGroup", ruleSrcGroupCount))
+
 	return &ruleSrcGroup, ruleSrcGroupPath, nil
 }
 
-func (service *SecurityPolicyService) updatePeerExpressions(obj *v1alpha1.SecurityPolicy, peer *v1alpha1.SecurityPolicyPeer, group *model.Group, idx int) {
+func (service *SecurityPolicyService) updatePeerExpressions(obj *v1alpha1.SecurityPolicy, peer *v1alpha1.SecurityPolicyPeer, group *model.Group, idx int) int {
 	service.appendOperatorIfNeeded(&group.Expression, "OR")
-	expressions := data.NewListValue()
-	expressionFrame := data.NewStructValue(
-		"",
-		map[string]data.DataValue{
-			"expressions":   expressions,
-			"resource_type": data.NewStringValue("NestedExpression"),
-		},
-	)
-	group.Expression = append(group.Expression, expressionFrame)
+	expressions := service.buildGroupExpression(&group.Expression)
 
 	clusterExpression := service.buildExpression(
 		"Condition", "SegmentPort",
 		fmt.Sprintf("%s|%s", util.TagScopeNCPCluster, service.getCluster()),
-		"Tag", "EQUALS",
+		"Tag", "EQUALS", "EQUALS",
 	)
 	expressions.Add(clusterExpression)
+	var err error = nil
+	var podSelectorCount, VMSelectorCount, nsSelectorCount = 0, 0, 0
 
 	if peer.PodSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		podExpression := service.buildExpression(
-			"Condition", "SegmentPort", util.TagScopeNCPPod, "Tag", "EQUALS")
+			"Condition", "SegmentPort", util.TagScopeNCPPod, "Tag", "EQUALS", "EQUALS")
 		expressions.Add(podExpression)
-		service.updatePortExpressions(peer.PodSelector.MatchLabels, expressions)
+
+		service.updateExpressionsMatchLables(peer.PodSelector.MatchLabels, "SegmentPort", expressions)
+		if len(peer.PodSelector.MatchLabels) > 0 {
+			podSelectorCount = 1
+		}
+
+		if peer.PodSelector.MatchExpressions != nil {
+			mergedMatchExpressions := service.mergeSelectorMatchExpression(peer.PodSelector.MatchExpressions)
+			podSelectorCount, err = service.validateSelectorExpressions(peer.PodSelector.MatchLabels, *mergedMatchExpressions, "SegmentPort")
```

Comment:
I feel not comfortable with validating while constructing a policy because that will introduce more validating during the process, which makes the code complex. Essentially we only want to know if the final policy/group/rule is legal, we don't care about in which step the exceeding occurs, right? So do you think if we have any way to validate only in the last step to simplify the code?

Reply:
After thought, I think we had better to do pre-check for operatorIn case and put nsx limit validation in the last step after expression build completed. Since operatorIn is a special case which is tightly connected to logic of the next step matchexpression build.  After separation , it can decouple logic for expression build and expression validation.